tracer/src/Datadog.Trace/Debugger/Symbols/SymbolExtractor.cs

```diff
@@ -0,0 +1,385 @@
+
+            var classLanguageSpecifics = new LanguageSpecifics
+            {
+                AccessModifiers = new[] { (type.Attributes & TypeAttributes.VisibilityMask).ToString() },
+                Annotations = new[] { type.Attributes.ToString() },
+                Interfaces = interfaceNames,
+                SuperClasses = baseClassNames
+            };
+            return classLanguageSpecifics;
+        }
+
+        internal Root GetAssemblySymbol(Assembly assembly, string serviceName)
+        {
+            var assemblyScope = new Model.Scope
+            {
+                Name = assembly.FullName,
+                ScopeType = SymbolType.Assembly,
+                SourceFile = assembly.Location,
+                StartLine = -1,
+                EndLine = int.MaxValue,
+                LanguageSpecifics = null,
+                Scopes = new List<Model.Scope>()
+            };
+
+            var root = new Root
+            {
+                Service = serviceName,
+                Env = string.Empty,
+                Language = "dotnet",
+                Version = string.Empty,
+                Scopes = new[] { assemblyScope }
+            };
+
+            return root;
+        }
+
+        private Symbol[] GetFieldSymbols(IList<FieldDef> typeFields)
+        {
+            if (typeFields.Count <= 0)
+            {
+                return null;
+            }
+
+            var typeSymbols = new Symbol[typeFields.Count];
+            for (var j = 0; j < typeSymbols.Length; j++)
+            {
+                var field = typeFields[j];
+                typeSymbols[j] = new Symbol
+                {
+                    Name = field.Name,
+                    SymbolType = field.IsStatic ? SymbolType.Static_Field : SymbolType.Field,
+                    Type = field.FieldType.TypeName,
+                    Line = -1
+                };
+            }
+
+            return typeSymbols;
+        }
+
+        private List<Model.Scope> GetMethodSymbols(IList<MethodDef> typeMethods, DatadogPdbReader pdbReader, out int classStartLine, out int classEndLine, out string typeSourceFile)
+        {
+            classStartLine = -1;
+            classEndLine = -1;
+            typeSourceFile = null;
+
+            if (typeMethods.Count <= 0)
+            {
+                return null;
+            }
+
+            var classScopes = ListCache<Model.Scope>.AllocList();
+
+            for (var j = 0; j < typeMethods.Count; j++)
+            {
+                var method = typeMethods[j];
+                var symbolMethod = pdbReader.GetMethodSymbolInfo(method.MDToken.ToInt32());
+                if (symbolMethod == null)
+                {
+                    continue;
+                }
+
+                var firstSq = symbolMethod.SequencePoints.FirstOrDefault(sq => sq.IsHidden() == false);
+                typeSourceFile ??= firstSq.Document?.URL;
+                var startLine = firstSq.Line == 0 ? -1 : firstSq.Line;
```

Comment:
-1 should be put in cases where we want to complete this symbol on any line in the file. For example fields in a class should be completed in all lines of all methods in that file.

Reply:
What about methods that we failed to determine their start line?

Reply:
Is this a common occurrence? I would consider this a bug. 🤔

Reply:
Well, for now - yes. The possibility it could happen is very slim since we are basing all our analysis on the existence of symbolic information.
In .NET, all the symbolic information is kept in a separate file `X.PDB`. These PDB files may not present in which case the best we will be able to do is use Reflection to extract whatever we can (Type hierarchy, Argument names, Method names and signature, etc) but won't be able to put our hands on more advanced information that is only kept in the PDB files: Local variable names, location of the methods in the original source file, etc...
In which case, my gut feeling is that we should still upload _something_ (the only information available we have). I'm wondering how the backend will handle it.

Reply:
There are two main use cases the backend serves:
1. IDE-like code completion in places where expression language can be used
2. Search-as-you type experience for class- and method names

With no line information on methods and variables I think serving use case 1. is difficult to impossible. Use case 2. should still be possible by uploading the class and method scopes without line information.

Reply:
So I think having only class names, the contained method names and their arguments will serve well the second use-case.
When PDB is missing we won't be able to serve the first use-case which is fine.